### PR TITLE
Allow CLI --options after parameters

### PR DESCRIPTION
**Pull request by kke**
Allow command options and flags after the command arguments.

Before:

```console
$ kontena plugin install shell --version 0.1.0
ERROR: too many arguments
$ kontena plugin install shell --help
ERROR: too many arguments
$ kontena plugin install --help shell
Usage:
    kontena plugin install [OPTIONS] NAME
...
```

After:

```console
$ kontena plugin install shell --version 0.1.0
[done] Installing plugin shell
Installed plugin shell version 0.1.0
$ kontena plugin install shell --help
Usage:
    kontena plugin install [OPTIONS] NAME
...
```

**Note:**

When using "remote exec commands" such as `kontena node ssh` or `kontena service exec`, you need to use the common double dash `--` escaping to separate remote command parameters from Kontena CLI parameters, for example:

```console
$ kontena service exec -it foo/bar -- ls -al
```

### CODE REVIEW

File: cli/lib/kontena/cli/containers/exec_command.rb [hunk 1]
```diff
@@ -6,6 +6,8 @@ class ExecCommand < Kontena::Command
     include Kontena::Cli::GridOptions
     include Kontena::Cli::Helpers::ExecHelper
 
+    Clamp.allow_options_after_parameters = false
```
Comment (SpComb) [inline cli/lib/kontena/cli/containers/exec_command.rb]: Move these assignments outside of the class to be less deceptive?

**Comment by kke** [comment]
Btw, even without this PR, something like: `kontena service exec foo/bar ls -D`will turn local debugging on and still pass the `-D` to target.

File: cli/lib/kontena/cli/containers/exec_command.rb (hunk 1 above)
Comment (kke) [reply cli/lib/kontena/cli/containers/exec_command.rb]: Tried something more fancy, yay or nay?

Comment (kke) [reply cli/lib/kontena/cli/containers/exec_command.rb]: Yet another solution

File: cli/lib/kontena/callback.rb [hunk 2]
```diff
@@ -22,14 +22,17 @@ def self.matches_commands(*commands)
       else
         cmd_type = cmd_type.to_sym
       end
-      cmd_types[cmd_class.to_sym] = cmd_type
+      cmd_types[cmd_class.to_sym] ||= []
```
Comment (kke) [inline cli/lib/kontena/callback.rb]: I have no clue why I made this double loop thing back then.

Anyways, the problem here was that without this fix, something like:
```
matches_commands '* exec', '* ssh'  
```
would only add the last one.

File: cli/lib/kontena/callbacks/exec_commands/option_placement.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+module Kontena
+  module Callbacks
+    class OptionPlacement < Kontena::Callback
+      matches_commands '* exec', '* ssh'
```
Comment (kke) [inline cli/lib/kontena/callbacks/exec_commands/option_placement.rb]: The upside of this solution is that plugin commands such as `kontena vagrant master ssh` are now handled too.

File: cli/lib/kontena/cli/containers/exec_command.rb (hunk 1 above)
Comment (SpComb) [reply cli/lib/kontena/cli/containers/exec_command.rb]: Relying on the subcommand loader would be fine, I just meant that the code should be clearer as to what's going on.

**Comment by SpComb** [comment]
An e2e spec with a new process per command would be preferable over a normal CLI spec in this case, because the code under test is touching global variables.

Alternatively I assume that clamp would understand something like `kontena node ssh -u foo test-node -- ls -la` so we could just enable it globally... but it's probably worth special-casing the `ssh` and `exec` commands, because you're more likely than not to run into issues otherwise.

**Comment by kke** [comment]
> Alternatively I assume that clamp would understand something like kontena node ssh -u foo test-node -- ls -la

Looks like it can:

```console
$ bin/kontena service exec -it foo/api ls -al
ERROR: Unrecognised option '-l'
$ bin/kontena service exec -it foo/api -- ls -al
total 192
drwxr-xr-x   11 root     root          4096 Feb  8 12:54 .
drwxr-xr-x    1 root     root          4096 Feb  8 12:55 ..
-rw-r--r--    2 root     root          2382 Feb  8 12:53 .drone.yml
```

Maybe get rid of the callback?

**Comment by SpComb** [comment]
Could we extend the `kontena node ssh --help` etc. usage info to include `--`? Right now it says `kontena node ssh [OPTIONS] [NODE] [COMMANDS] ...`, it could include the `--` as a reminder/example, aka

```
Usage:
    kontena node ssh [OPTIONS] [NODE] -- [COMMANDS] ...
```

File: test/spec/features/cli_spec.rb [hunk 4]
```diff
@@ -8,4 +8,22 @@
       expect(k.out).to match(/Options:/)
     end
   end
+
+  context 'option parsing' do
+    after do
+      k.run 'kontena vault rm --force testsecret'
```
Comment (SpComb) [inline test/spec/features/cli_spec.rb]:
```
  1) cli option parsing allows options after parameters
     Failure/Error: k.run 'kontena vault rm --force testsecret'
     NameError:
       undefined local variable or method `k' for #<RSpec::ExampleGroups::Cli::OptionParsing:0x000000015a7458>
     # ./spec/features/cli_spec.rb:14:in `block (3 levels) in <top (required)>'
```

**Comment by kke** [comment]
> Could we extend the kontena node ssh --help etc. usage info to include --?

Was easier than expected

**Comment by SpComb** [comment]
One remaining special case: the `kontena node/master ssh` => `kontena vagrant ...` needs to also include the `--` now:

```
$ DEBUG=true kontena node ssh core-01 -- ls -la
DEBUG Kontena CLI 1.5.0.dev (ruby-2.3.1+x86_64-linux-gnu)
DEBUG Loading plugin kontena-plugin-cloud from /home/kontena/.kontena/gems/2.3.1/gems/kontena-plugin-cloud-1.2.0/lib/kontena_cli_plugin.rb
DEBUG Loading plugin kontena-plugin-digitalocean from /home/kontena/.kontena/gems/2.3.1/gems/kontena-plugin-digitalocean-0.3.3/lib/kontena_cli_plugin.rb
DEBUG Loading plugin kontena-plugin-vagrant from /home/kontena/.kontena/gems/2.3.1/gems/kontena-plugin-vagrant-0.3.1/lib/kontena_cli_plugin.rb
DEBUG Running Kontena::MainCommand with ["node", "ssh", "core-01", "--", "ls", "-la"] -- callback matcher = 'nil'
DEBUG Running Kontena::Cli::NodeCommand with ["ssh", "core-01", "--", "ls", "-la"] -- callback matcher = 'nil'
DEBUG Loading configuration from /home/kontena/.kontena_client.json
DEBUG Configuration loaded with 5 servers and 2 accounts.
DEBUG Current master: development
DEBUG Current grid: development
DEBUG Current account: kontena
DEBUG Running Kontena::Cli::Nodes::SshCommand with ["core-01", "--", "ls", "-la"] -- callback matcher = 'nodes ssh'
DEBUG Excon opts: {:omit_default_port=>true, :connect_timeout=>10, :read_timeout=>30, :write_timeout=>10, :ssl_verify_peer=>true, :middlewares=>[Excon::Middleware::ResponseParser, Excon::Middleware::Expects, Excon::Middleware::Idempotent, Excon::Middleware::Instrumentor, Excon::Middleware::Mock, Excon::Middleware::Decompress], :instrumentor=>Kontena::DebugInstrumentor, :ssl_ca_file=>nil, :ssl_verify_peer_host=>nil}
DEBUG [Request]: GET http://localhost:9292/v1/nodes/development/core-01 | Headers: {Accept: application/json, Accept-Encoding: gzip, Authorization: Bearer} 
DEBUG [Response]: Headers: {Content-Type: application/json, X-Kontena-Version: 1.5.0.dev}  | Status: 200 | Body: "{\"id\":\"development/core-01\",\"node_id\":\"XI4K:NPOL:EQJ4:S4V7:EN3B:DHC5:KZJ..."
DEBUG Running Kontena.run(["vagrant", "node", "ssh", "core-01", "ls", "-la"])
DEBUG Running Kontena::MainCommand with ["vagrant", "node", "ssh", "core-01", "ls", "-la"] -- callback matcher = 'nil'
DEBUG Running Kontena::Plugin::VagrantCommand with ["node", "ssh", "core-01", "ls", "-la"] -- callback matcher = 'nil'
DEBUG Running Kontena::Plugin::Vagrant::NodeCommand with ["ssh", "core-01", "ls", "-la"] -- callback matcher = 'nil'
DEBUG Running Kontena::Plugin::Vagrant::Nodes::SshCommand with ["core-01", "ls", "-la"] -- callback matcher = 'nodes ssh'
ERROR Command [["vagrant", "node", "ssh", "core-01", "ls", "-la"]] exception
ERROR Unrecognised option '-l' (Clamp::UsageError)
/home/kontena/.gem/ruby/2.3.0/gems/clamp-1.2.1/lib/clamp/option/parsing.rb:78:in `find_option'
/home/kontena/.gem/ruby/2.3.0/gems/clamp-1.2.1/lib/clamp/option/parsing.rb:41:in `set_options_from_command_line'
/home/kontena/.gem/ruby/2.3.0/gems/clamp-1.2.1/lib/clamp/option/parsing.rb:14:in `parse_options'
/home/kontena/.gem/ruby/2.3.0/gems/clamp-1.2.1/lib/clamp/command.rb:48:in `parse'
/home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.5.0.dev/lib/kontena/command.rb:206:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.5.0.dev/lib/kontena/command.rb:216:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.5.0.dev/lib/kontena/command.rb:216:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.5.0.dev/lib/kontena/command.rb:216:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.5.0.dev/lib/kontena_cli.rb:41:in `run!'
/home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.5.0.dev/lib/kontena/cli/nodes/ssh_command.rb:41:in `execute'
/home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.5.0.dev/lib/kontena/command.rb:216:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.5.0.dev/lib/kontena/command.rb:216:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/clamp-1.2.1/lib/clamp/subcommand/execution.rb:11:in `execute'
/home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.5.0.dev/lib/kontena/command.rb:216:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/clamp-1.2.1/lib/clamp/command.rb:132:in `run'
/home/kontena/.gem/ruby/2.3.0/gems/kontena-cli-1.5.0.dev/bin/kontena:22:in `<top (required)>'
/home/kontena/.local/bin/kontena:23:in `load'
/home/kontena/.local/bin/kontena:23:in `<main>'
ERROR: Unrecognised option '-l'

See: 'kontena_cli.rb vagrant node ssh --help'
DEBUG Execution took 0.319 seconds
DEBUG SystemExit status 1
```

**Pull request merged**